packages/sdk/ts/trees/stateTree.ts

```diff
@@ -65,3 +62,69 @@ export const generateSignUpTree = async ({
     publicKeys,
   };
 };
+
+/**
+ * Generate a State tree object from the events of a MACI smart contracts
+ * @param provider - the ethereum provider
+ * @param address - the address of the MACI contract
+ * @param fromBlock - the block number from which to start fetching events
+ * @param blocksPerRequest - the number of blocks to fetch in each request
+ * @param endBlock - the block number at which to stop fetching events
+ * @param sleepAmount - the amount of time to sleep between each request
+ * @returns State tree
+ */
+export const generateSignUpTreeWithEndKey = async ({
+  provider,
+  address,
+  fromBlock = 0,
+  blocksPerRequest = 50,
+  endBlock,
+  sleepAmount,
+  userPublicKey,
+}: IGenerateSignUpTreeWithEndKeyArgs): Promise<IGenerateSignUpTree> => {
+  const lastBlock = endBlock || (await provider.getBlockNumber());
+
+  const maciContract = MACIFactory.connect(address, provider);
+  const signUpTree = new LeanIMT(hashLeanIMT as LeanIMTHashFunction);
+  signUpTree.insert(PAD_KEY_HASH);
+  const publicKeys: PublicKey[] = [];
+
+  // Fetch event logs in batches (lastBlock inclusive)
+  for (let i = fromBlock; i <= lastBlock; i += blocksPerRequest + 1) {
+    // the last block batch will be either current iteration block + blockPerRequest
+    // or the end block if it is set
+    const toBlock = i + blocksPerRequest >= lastBlock ? lastBlock : i + blocksPerRequest;
+
+    // eslint-disable-next-line no-await-in-loop
+    const signUpLogs = await maciContract.queryFilter(maciContract.filters.SignUp(), i, toBlock);
+    // eslint-disable-next-line @typescript-eslint/prefer-for-of
+    for (let j = 0; j < signUpLogs.length; j += 1) {
+      const event = signUpLogs[j];
+      const publicKeyX = event.args._userPublicKeyX;
+      const publicKeyY = event.args._userPublicKeyY;
+
+      const publicKey = new PublicKey([publicKeyX, publicKeyY]);
+
+      publicKeys.push(publicKey);
+      signUpTree.insert(hashLeftRight(publicKeyX, publicKeyY));
+
+      // early return cause we found the user
+      if (publicKey.equals(userPublicKey)) {
+        return {
+          signUpTree,
+          publicKeys,
+        };
+      }
+    }
+
+    if (sleepAmount) {
+      // eslint-disable-next-line no-await-in-loop
+      await sleep(sleepAmount);
+    }
```

Comment:
Why do you need `sleep` here?

Reply:
we always had this feature when fetching large amount of data from RPCs, so that it can help bypass rate limiting in case of using free endpoints